### PR TITLE
Verify twin signatures for Edge Agent and Edge Hub

### PR DESCRIPTION
**Pull request by ggjjj**
1. Added Verify signatures function as util function to be shared by modules
2. Added test cases for Edge Agent & Edge Hub to verify signatures

### CODE REVIEW

File: Microsoft.Azure.Devices.Edge.sln [hunk 1]
```diff
@@ -217,6 +217,10 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EdgeHubTriggerCSharp", "edg
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "NumberLogger", "test\modules\NumberLogger\NumberLogger.csproj", "{B0151FFD-85C4-42B9-81CC-9193BADFE9F8}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "jsoncanonicalizer", "edge-util\src\jsoncanonicalizer\jsoncanonicalizer.csproj", "{1915CE7E-E949-4633-9862-D92BD5E873B5}"
```
Comment (CindyXing) [inline Microsoft.Azure.Devices.Edge.sln:220]: Can we keep this copy and let tools referencing source code at edgeutil? This way, we don't need to keep the same reference code in two places.

**Review by CindyXing — CHANGES_REQUESTED** [review]

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs [hunk 2]
```diff
@@ -395,6 +420,41 @@ Task UpdateDeploymentConfig(TwinCollection desiredProperties)
...
+        public static bool ExtractAgentTwinAndVerify(TwinCollection twinDesiredProperties)
+        {
+            try
+            {
+                // Extract Desired properties
+                JObject desiredProperties = new JObject();
+                JObject twinJobject = JObject.Parse(twinDesiredProperties.ToString());
+                desiredProperties["modules"] = twinJobject["modules"];
+                desiredProperties["runtime"] = twinJobject["runtime"];
+                desiredProperties["schemaVersion"] = twinJobject["schemaVersion"];
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:432]: schemaVersion is different from manifest version. It has always remained 1.0. The manifest version is right now included under the integrity section.

[the remark this answers is not in the document]
Comment (ggjjj) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:432]: Example in test case. https://github.com/Azure/iotedge/blob/154b1d95a8ec21ae2d7ed6cb1867deacc7453c25/edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.IoTHub.Test/EdgeAgentConnectionTest.cs#L1574

File: Microsoft.Azure.Devices.Edge.sln (hunk 1 above)
Comment (CindyXing) [reply Microsoft.Azure.Devices.Edge.sln:220]: I didn't mean the solution file.... 

I meant we are keeping the same source code under both edgeutil and tools\manifestsigner folder. 
Instead we can just keep the one under edgeutil; and have tools\manifestsigner reference it.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:432]: Ok. Let's discuss the manifest version.

**Comment by ggjjj** [comment]
Closing this PR as addressing in an another PR and sending merging req to iiot branch

**Pull request closed without merging**